docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+# Architecture
+
+The core of OSP-Magnum is best described as a library that provides a set of functions and data structures intended to support a space flight simulator. The user is responsible for implementing their own application on the top level, allowing for full control over every feature.
+
+OSP-Magnum provides two distinct worlds the application can manage:
+* **Universe**: A space simulator that can handle massive world sizes with multiple coordinate spaces. Objects in space are referred to as **Satellites**.
+* **ActiveScene**: A typical game engine that can support physics, rendering, etc..
+
+The ActiveScene can be synchronized to represent a small area of the Universe, where in-game representations of nearby Satellites can be added to the scene. This area is referred to as the **ActiveArea** and can follow the user around to seamlessly explore the Universe while also avoiding floating point precision errors.
+
+For performance and flexibility for such a project, we adhere to few techniques that may seem unconventional:
+* Data-oriented Design
+* Entity Component System
+
+## Data-oriented Design (DoD)
+
+Data-oriented Design is an emerging approach to software architecture, used extensively in OSP-Magnum. By focusing on creating a memory layout for efficient access, DoD greatly reduces complexity and achieves high performance. As counterintuitive as it may seem, this potentially results in (or at least encourages) a greater degree of flexibility compared to class-based OOP.
+
+DoD is not a new idea, but it is incredibly unconventional to most programmers (especially ones adept to OOP). It will be useful to read about or listen to a few presentations before getting started.
+
+See [data-oriented-design](https://github.com/dbartolini/data-oriented-design) for plenty of resources.
+
+Important note: When going down the rabbit hole, *Data-oriented programming* and *Data-driven design* is not what you're looking for. *Correct* resources on DoD usually mention the CPU cache on systems-level languages.
+
+*tl;dr: RAM sucks at random access and use arrays everywhere because they're fast!*
+
+### DoD in practice
+
+DoD leads to:
+
+* **Seprated Data and Functions**. This idea is also favored by functional programming. Most classes are only responsible for storing data and their methods only modify their own members. Most operations are free functions that only have a few responsibilities and little side effects.
+* **Structure of Arrays**. Aside from efficiency, data can be organized based on how it's operated on. Data for a single 'object' can be represented across multiple arrays, and each of these arrays can be passed to functions that can run simultaneously. This also leads to a natural separation of concerns.
```

Comment:
"Separating a game object's state data (position, velocity, etc.) into standalone components improves memory performance. By collecting all components of a particular type (e.g. position) into a contiguous array, a function which needs to use only that type of component is able to quickly read the data from start to finish without needing to jump over extraneous data (e.g. velocity), as it would if these properties were packaged together. This arrangement also lends itself to a natural separation of concerns, since functions need not know about any data that is not relevant to them."